zoneinfo/src/types.rs

```diff
@@ -0,0 +1,650 @@
+            Ok(Self::Numeric)
+        } else if s.contains("/") {
+            let (std, dst) = s.split_once('/').ok_or(ZoneInfoParseError::UnknownValue(
+                ctx.line_number,
+                s.to_owned(),
+            ))?;
+            Ok(Self::Pair(std.to_owned(), dst.to_owned()))
+        } else {
+            Ok(AbbreviationFormat::String(s.to_owned()))
+        };
+        ctx.exit();
+        value
+    }
+}
+
+impl Default for AbbreviationFormat {
+    fn default() -> Self {
+        Self::String("LMT".to_owned())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct FormattableAbbr(String);
+
+impl FormattableAbbr {
+    pub fn to_formatted_string(&self, letter: &str) -> String {
+        self.0.replace("%s", letter)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct UntilDateTime {
+    pub date: Date,
+    pub time: QualifiedTime,
+}
+
+impl UntilDateTime {
+    pub fn as_date_secs(self) -> i64 {
+        self.date.as_secs()
+    }
+
+    pub fn as_precise_ut_time(self, std_offset: i64, save: i64) -> i64 {
+        self.as_date_secs() + self.time.to_universal_seconds(std_offset, save)
+    }
+}
+
+impl TryFromStr<LineParseContext> for UntilDateTime {
+    type Error = ZoneInfoParseError;
+    fn try_from_str(s: &str, ctx: &mut LineParseContext) -> Result<Self, Self::Error> {
+        ctx.enter("UntilDateTime");
+        let mut splits = s.split_whitespace();
+        let year = next_split(&mut splits, ctx)?.context_parse::<i32>(ctx)?;
+        let date_or_end = splits.next();
+        let date = if let Some(month) = date_or_end {
+            let month = month.context_parse::<Month>(ctx)?;
+            let day = next_split(&mut splits, ctx)
+                .ok()
+                .map(|s| s.context_parse::<DayOfMonth>(ctx))
+                .transpose()?
+                .unwrap_or(DayOfMonth::Day(1));
+            Date { year, month, day }
+        } else {
+            ctx.exit();
+            return Ok(UntilDateTime {
+                date: Date {
+                    year,
+                    month: Month::Jan,
+                    day: DayOfMonth::Day(1),
+                },
+                time: QualifiedTime::Local(Time::default()),
+            });
+        };
+
+        let time = next_split(&mut splits, ctx)
+            .ok()
+            .map(|t| t.context_parse::<QualifiedTime>(ctx))
+            .transpose()?
+            .unwrap_or(QualifiedTime::Local(Time::default()));
+
+        ctx.exit();
+        Ok(Self { date, time })
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct Date {
+    pub year: i32,
+    pub month: Month,
+    pub day: DayOfMonth,
+}
+
+impl Date {
+    pub fn as_secs(&self) -> i64 {
+        let epoch_days = epoch_days_for_rule_date(self.year, self.month, self.day);
+        utils::epoch_seconds_for_epoch_days(epoch_days)
+    }
+}
+
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
+pub struct Time {
+    pub sign: Sign,
```

Comment:
question: The significance of `sign` isn't clear for a type named `Time`. Does this represent a length of time or something else?

Reply:
The `Time` type here is pretty broad and is essentially anything that is `[+|-]hh:mm:ss`, so I guess it depends on the context.

Typically, it's an offset amount or savings amount, while `QualifiedTime` is more of a proper "instant".